src/VisualStudio/Core/Def/Implementation/UnusedReferences/ProjectAssets/ProjectAssetsReader.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis.UnusedReferences;
+using Newtonsoft.Json;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.UnusedReferences.ProjectAssets
+{
+    // This class will read the dependency heirarchy from the project.assets.json file. The format of this file
+    // is subject to change and in the future this information will be provided by an API.  See https://github.com/dotnet/roslyn/issues/50054
+    internal static class ProjectAssetsReader
+    {
+        // NuGet will include entries to keep empty folders from being removed. These entries can be ignored.
+        private const string NuGetEmptyFileName = "_._";
+
+        public static ImmutableArray<ReferenceInfo> ReadReferences(
+            ImmutableArray<ReferenceInfo> projectReferences,
+            string projectAssetsFilePath,
+            string targetFrameworkMoniker)
+        {
+            var projectAssetsFileContents = File.ReadAllText(projectAssetsFilePath);
+            var projectAssets = JsonConvert.DeserializeObject<ProjectAssetsFile>(projectAssetsFileContents);
+
+            if (projectAssets is null ||
+                projectAssets.Version != 3)
+            {
+                return ImmutableArray<ReferenceInfo>.Empty;
+            }
+
+            if (projectAssets.Targets is null ||
+                !projectAssets.Targets.TryGetValue(targetFrameworkMoniker, out var target))
+            {
+                return ImmutableArray<ReferenceInfo>.Empty;
+            }
+
+            var autoReferences = projectAssets.Project?.Frameworks?.Values
+                .SelectMany(framework => framework.Dependencies?.Keys.Where(key => framework.Dependencies[key].AutoReferenced))
+                .Distinct()
+                .ToImmutableHashSet();
+            autoReferences ??= ImmutableHashSet<string>.Empty;
+
+            var references = projectReferences
+                .Select(projectReference => BuildReference(projectAssets, target, projectReference, autoReferences))
+                .OfType<ReferenceInfo>()
+                .ToImmutableArray();
+
+            return references;
+        }
+
+        private static ReferenceInfo? BuildReference(
+            ProjectAssetsFile projectAssets,
+            Dictionary<string, ProjectAssetsTargetLibrary> target,
+            ReferenceInfo projectReference,
+            ImmutableHashSet<string> autoReferences)
+        {
+            var referenceName = projectReference.ReferenceType == ReferenceType.Project
```

Comment:
I admit calling the local variable here 'projectReference' is confusing, since this line then looks like "is the project reference a project reference?"